src/interpreter/type_cast.rs

```diff
@@ -12,138 +12,56 @@
-        if let Some(number) = value.downcast_ref::<u8>() {
-            return Ok(PactType::Numeric(Numeric(*number as u64)));
-        }
-        if let Some(number) = value.downcast_ref::<u16>() {
-            return Ok(PactType::Numeric(Numeric(*number as u64)));
-        }
-        if let Some(number) = value.downcast_ref::<u32>() {
-            return Ok(PactType::Numeric(Numeric(*number as u64)));
-        }
-        if let Some(number) = value.downcast_ref::<u64>() {
-            return Ok(PactType::Numeric(Numeric(*number)));
-        }
-        if let Some(number) = value.downcast_ref::<u128>() {
-            return Ok(PactType::Numeric(Numeric(
-                u64::try_from(*number).map_err(|_| PactConversionErr::Overflow)?,
-            )));
-        }
-
-        // String-like type casting into PactType
-        if let Some(string) = value.downcast_ref::<&str>() {
-            return Ok(PactType::StringLike(StringLike(&*string.as_bytes())));
-        }
-        if let Some(string) = value.downcast_ref::<String>() {
-            return Ok(PactType::StringLike(StringLike(string.as_bytes())));
-        }
-        if let Some(string) = value.downcast_ref::<Vec<u8>>() {
-            return Ok(PactType::StringLike(StringLike(&*string)));
-        }
+/// Dummy structs
+struct Numbers;
+struct Strings;
 
-        // Fixed hash type casting into PactType
-        if let Some(string) = value.downcast_ref::<[u8; 4]>() {
-            return Ok(PactType::StringLike(StringLike(&*string))); // H32
-        }
-        if let Some(string) = value.downcast_ref::<[u8; 8]>() {
-            return Ok(PactType::StringLike(StringLike(&*string))); // H64
-        }
-        if let Some(string) = value.downcast_ref::<[u8; 16]>() {
-            return Ok(PactType::StringLike(StringLike(&*string))); // H128
-        }
-        if let Some(string) = value.downcast_ref::<[u8; 20]>() {
-            return Ok(PactType::StringLike(StringLike(&*string))); // H160
-        }
-        if let Some(string) = value.downcast_ref::<[u8; 32]>() {
-            return Ok(PactType::StringLike(StringLike(&*string))); // H256
-        }
-        if let Some(string) = value.downcast_ref::<[u8; 33]>() {
-            return Ok(PactType::StringLike(StringLike(&*string))); // H264
-        }
-        if let Some(string) = value.downcast_ref::<[u8; 64]>() {
-            return Ok(PactType::StringLike(StringLike(&*string))); // H512
-        }
-        if let Some(string) = value.downcast_ref::<[u8; 65]>() {
-            return Ok(PactType::StringLike(StringLike(&*string))); // H520
-        }
-        if let Some(string) = value.downcast_ref::<[u8; 128]>() {
-            return Ok(PactType::StringLike(StringLike(&*string))); // H1024
-        }
-        if let Some(string) = value.downcast_ref::<[u8; 256]>() {
-            return Ok(PactType::StringLike(StringLike(&*string))); // H2048
-        }
+/// Impl for all types that implement lossless conversion into u64
```

Comment:
We still want to support `u128` here. Will have to wait until #1 is implemented before we can make `Numeric(u128)` and simplify it like this